src/Backend/Haskell.idr

```diff
@@ -358,25 +369,25 @@ mutual
   makeDefs    (TSum xs)   = concat <$> traverse (assert_total makeDefs) xs
   makeDefs    (TVar v)    = pure []
   makeDefs td@(TMu cases) = makeDefs' $ TName (nameMu cases) td -- We name anonymous mus using their constructors.
-  makeDefs    (TApp f xs) = do
-      res <- assert_total $ makeDefs' f
-      res' <- concat <$> traverse (assert_total makeDefs) xs
-      pure (res ++ res')
+  makeDefs    (TApp f xs) = 
+    do res <- assert_total $ makeDefs' f
+       res' <- concat <$> traverse (assert_total makeDefs) xs
+       pure (res ++ res')
 
   ||| Generate Haskell type definitions for a `TNamed` and all of its dependencies.
   makeDefs' : TNamed n -> State (List Name) (List Haskell)
   makeDefs' (TName name body) = ifNotPresent name $
       let freshEnvString = map (\ x => case x of HsVar v => v; _ => "") freshEnv
           decl = MkDecl name (getUsedVars freshEnvString body) in
       case body of
-        TMu cases => do -- Named `TMu`s are treated as ADTs.
-          let newEnv = hsParam decl :: freshEnv
-          let args = map (map (makeType newEnv)) cases
-          res <- map concat $ traverse {b=List Haskell} (\(_, bdy) => assert_total $ makeDefs bdy) (toList cases)
-          pure $ ADT decl args :: res
-        _ => do -- All other named types are treated as synonyms.
-          res <- assert_total $ makeDefs body
-          pure $ Synonym decl (makeType freshEnv body) :: res
+        TMu cases => -- Named `TMu`s are treated as ADTs.
+          do let newEnv = hsParam decl :: freshEnv
+             let args = map (map (makeType newEnv)) cases
+             res <- map concat $ traverse {b=List Haskell} (\(_, bdy) => assert_total $ makeDefs bdy) (toList cases)
+             pure $ res ++ [ADT decl args]
```

Comment:
Since we're generating things in the order they were introduced, `reverse` in the code generator is not needed anymore. However when generating things for nested definitions, this means you'd get the output for your thing before its dependencies if you append. Prepending solves this.